File: day2/main.js
const fs = require('fs');

const MAX_RED = 12;
const MAX_GREEN = 13;
const MAX_BLUE = 14;

main();

function main()
{
    let data;
    let result = 0;
    let sum = 0;

    try 
    {
        data = fs.readFileSync('./input.txt', 'utf8');
    } 
    catch (err) 
    {
        throw new Error("Error reading the file");
    }

    let fileLineByLine = data.split(/\n/);

    for (let index = 0; index < fileLineByLine.length; index++) 
    {
        if (evaluateGame(fileLineByLine[index])) 
        {
            result = result + getGameNumber(fileLineByLine[index]);
        }
    }

    console.log(bagContent(fileLineByLine[0]));
    console.log(powerOfSet(bagContent(fileLineByLine[0])));

    for (let i = 0; i < fileLineByLine.length; i++)
    {
        sum = sum + powerOfSet(bagContent(fileLineByLine[i]));
    }

    console.log("Your result is: " + result);
    console.log("Sum of all powers of sets is: " + sum);
}

function powerOfSet(set)
{
    let result = 1;

    for (let i = 0; i < set.length; i++)
    {
        result = result * set[i];
    }

    return result;
}

function bagContent(game) 
{

    let draws = game.split(": ")[1].split(/; |, /);

    let draw;

    let reds = 0;
    let greens = 0;
    let blues = 0;

    for (let i = 0; i < draws.length; i++)
    {
        draw = draws[i].split(" ");

        switch (draw[1]) {
            case "red":
                reds = Math.max(reds, Number(draw[0]));
                break;
            case "green":
                greens = Math.max(greens, Number(draw[0]));
                break;
            case "blue":
                blues = Math.max(blues, Number(draw[0]));
                break;
        
            default:
                break;
        }
    }

    return [reds, greens, blues];
}

//This function return true when game is possible, and false when not
function evaluateGame(game) 
{

    let draws = game.split(": ")[1].split(/; |, /);

    let draw;

    let reds = 0;
    let greens = 0;
    let blues = 0;

    for (let i = 0; i < draws.length; i++)
    {
        draw = draws[i].split(" ");

        switch (draw[1]) {
            case "red":
                reds = Math.max(reds, Number(draw[0]));
                break;
            case "green":
                greens = Math.max(greens, Number(draw[0]));
                break;
            case "blue":
                blues = Math.max(blues, Number(draw[0]));
                break;
        
            default:
                break;
        }
    }

    if (reds > MAX_RED || greens > MAX_GREEN || blues > MAX_BLUE)
    {
        return false;
    } 
    else
    {
        return true;
    }
}

//This function returns number of Game
function getGameNumber(game)
{
    let gameName = game.split(": ")[0];
    let gameNumber = gameName.replace(/Game /, '');

    return Number(gameNumber);
}

File: day2/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LINE_LENGHT 200
#define MAX_RED 12
#define MAX_GREEN 13
#define MAX_BLUE 14

int evaluateGame(char* line);
int getGameNumber(char* line);

int main(int argc, char const *argv[])
{
    FILE* filePointer = fopen("./testinput.txt", "r");
    char currentLine[LINE_LENGHT];

    int result;

    if (filePointer == NULL)
    {
        printf("ERROR: Could not open file.\n");
        return EXIT_FAILURE;
    }

    while (fgets(currentLine, LINE_LENGHT, filePointer))
    {
        result += evaluateGame(currentLine);
    }

    fclose(filePointer);

    printf("Program executed successfully. Result: %d\n", result);
    return EXIT_SUCCESS;
}

int evaluateGame(char* line)
{
    int gameNumber = getGameNumber(line);

    return gameNumber;
}

int evaluateGame(char* line)
{
    int index = 5;

    while (line[index] != 58) //58 is ASCII for ":"
    {
        index++;
    }
    
    for (index--; index > 4; index--)
    {
        
    }
}
